Wireless controller for machinery connections

Loganbacca opened the pull request:
This is a controller and framework for managing connections between various objects (i.e. buttons, doors, vents, mass drivers, etc.).

Currently to achieve this either the radio controller is used, or specific code is written into things like buttons that have ugly "world" loops and the like (emitter buttons for example).

All the wireless manager has to do in this instance is pass an initial connection request from a wireless sender onto a wireless receiver. From there the sender calls the receivers procs directly. 

To Do:
 - [x] Documentation
 - [x] Safety checks

Conversions:
 - [x] Emitters
 - [x] Crematorium
 - [x] Holosign
 - [x] Flasher
 - [x] Mass driver
 - [x] Igniter
 - [x] Blast doors

Feel free to comment on this implementation, I've never built a controller before so feedback on that would be appreciated. Also please suggest sanity checks and features/functionality.

I'll try to convert as many devices over as possible, as this new system should significantly reduce the processing overhead associated with remote devices.

EDIT: this PR now also includes a little helper datum for synchronizing multiple spawn() calls. See spawn_sync.dm for details.

This new system will run in parallel to the existing systems. I'll do a separate PR for updating the map.

PsiOmegaDelta commented:
If the wifi_manager is a global, singleton object it may be best to add all its functionality into the ```/datum/controller/process/wireless``` directly. Existing managers (event, nano, alarm, etc.) are there mostly because of legacy reasons.

The main reason why is because you want to be able to make a scheck() call as the last line of every for-each loop, to ask the scheduler if it's time to sleep and let other processes work which further helps with processing issues.

And since you are redoing how object communicate with each other, consider having a plan for allowing multitools to connect them. Don't necessarily have to implement it, just try to design things that adding the functionality later won't be too difficult.

Loganbacca commented:
Yep, the plan is to include functionality to connect devices in game. 

I'll modify the wireless controller to make it a proper process scheduler controller.

Hubblenaut commented:
Will this use an extra tool? It'd be nice if you will be able to edit connections using the already existing multitool to give it more purpose other than pulsing, fireaxe-locker-opening and t-comms-calibrating.

Loganbacca commented:
@Hubblenaut I'll get the functionality added to the multitool. Probably just at a basic level, and leave it to @PsiOmegaDelta to integrate properly seeing how he's doing interesting things with the multitool in #11458

Loganbacca commented:
I've added some more GC code - let me know if I'm missing anything.

I also got the base remote buttons working with Emitters and Crematoriums. From here it'll be easy to do the rest of the button activate objects, and they'll be able to be reconfigured in game.

PsiOmegaDelta commented:
The Destroy() code looks fine now, the null-checks you're making are optional tho as qdel() makes its own null-check anyway.

Loganbacca commented:
I've converted a few more things to use the wireless system. I also cleaned up mounted sparkers a little bit.

I also wanted synchronized door openings without things like mass drivers firing early, so I built a helper datum to help synchronize spawn()ing code. In short it will allow you to spawn() multiple calls, then "pause" your code with a while loop until all the spawns() have completed. Have a look under spawn_sync.dm if you want more info. I've currently implemented it in ```code/modules/wireless/devices.dm``` for the door opening and mass driver sequences.

I still need to document the wireless system properly, but otherwise this PR is pretty much feature complete.

Loganbacca commented:
All done and ready for review.

mwerezak commented on code/controllers/Processes/wireless.dm:
```diff
@@ -0,0 +1,73 @@
+		//any that fail are moved to the retry queue
+		process_queue(pending_connections, retry_connections)
+		//quit here so the retry connection attempt is delayed one cycle
+		return
```
Won't this mean that if new requests are coming in frequently enough, retry attempts will be delayed indefinitely?

mwerezak replied:
Or more simply, just reverse the order that they are processed:
```
if(retry_connections.len > 0)
    process_queue(retry_connections, failed_connections)

if(pending_connections.len > 0)
    process_queue(pending_connections, delayed)
```

Loganbacca replied:
I think my intention was to give new connections a priority, and leave the retries one cycle before trying them again. Swapping the order makes way more sense though, thanks for the suggestion.

mwerezak commented on code/controllers/Processes/wireless.dm:
```diff
@@ -0,0 +1,73 @@
+		//any that fail are moved into the failed connections list
+		process_queue(retry_connections, failed_connections)
+
+/datum/controller/process/wireless/proc/process_queue(var/list/input_queue, var/list/output_queue)
```
Minor, but it might be better for clarity to rename these parameters something like `process_connections` and `unsuccessful_connections`.

mwerezak commented on code/_helpers/spawn_sync.dm:
```diff
@@ -0,0 +1,75 @@
+//-------------------------------
+/datum/spawn_sync
+	var/count = 1
+	var/safety = 100	//aprox 10 seconds
```
Again minor, but it seems like you mean to release all of the semaphore's resources after 100 ticks, but the behaviour that is implemented is to release everything after 100 calls to `check()`. Would it be worthwhile to implement something based on `time` instead? It's a bit nicer that way since the implementation is more self-contained - it still works as intended even if client code does not call `check()` at regular intervals.

Perhaps less minor, if on the other hand client code should *always* use the `spawn_sync` semaphore as you have indicated in the comment above, then the boilerplate should probably be wrapped into helper methods. Maybe something like:
```
/datum/spawn_sync/proc/open(var/thread_proc, var/list/thread_args = list())
	//dispatch a new thread
	spawn()
		//Utilise try/catch keywords here so the code continues even if an error occurs.
		try
			//code to do stuff goes here
			call(thread_proc)(arglist(thread_args))	
		catch(var/e)
			sync.close()
			throw e //Runtime errors in SS13 are almost always unwanted, hopefully this will ensure the runtime is logged.

		sync.close()


/datum/spawn_sync/proc/block_until_done()
	finalize()

	//Create a while loop to check if the sync is complete yet, it will return true once all the spawn threads have 
	// completed.
	while(check())
		//Add a sleep call to delay each check.
		sleep(1)

```
This way client code can be more agnostic about the details of how `spawn_sync` works and just use it.

Loganbacca replied:
I like this, good idea. I've never used the call() proc, so this will be good practice.

mwerezak commented on code/game/machinery/buttons.dm:
```diff
@@ -5,14 +5,114 @@
 	idle_power_usage = 2
 	active_power_usage = 4
+	var/_wifi_id
+	var/_wifi_toggle = 0		//set this to 1 if you want the button to have both an on and off state, not just fire a single command
```
I'm kind of inclined to see toggle buttons be a subtype of one-off buttons (with an override of `activate()` instead of a switch off of the `_wifi_toggle` var), but maybe that's just taste.

Loganbacca replied:
Originally I made toggle switches a subtype. I'll go back and have a look at making them subtypes again.

mwerezak commented:
Pretty cool

Loganbacca commented:
@mwerezak thanks for all the input, I'll go through and implement those suggestions. 

And yes, it should be easy to add multitool functionality.

Loganbacca commented:
Code updated taking in suggestions by @mwerezak
Only thing I haven't touched is the toggle mode. I may go back and do that though.

Also added a changelog.

mwerezak commented on code/_helpers/spawn_sync.dm:
```diff
@@ -0,0 +1,91 @@
+
+//Failsafe in case something breaks horribly
+/datum/spawn_sync/proc/safety_check()
+	if(TimeOfHour > timer)
```
TimeOfHour rolling over is a possible concern. 

Now that spawn_sync is fully in control of the check() calls using a decrementing counter as you were doing earlier is perfectly fine.

Loganbacca replied:
I was actually wondering about whether TimeOfHour would have any rollover issues. I'll trim it out and see what I can do about the button toggle mode at the same time.

Loganbacca commented:
All done, unless there are anymore issues or suggestions?

mwerezak commented:
Should wifi senders and receivers have any support for checking the powered/broken/emped state of their host machine or should we always leave that responsibility to the machine?

Loganbacca commented:
IMO it should be up to the machine. I think it would just add unnecessary clutter to the wifi datums that they don't really need. Machines should really be checking their state before doing anything anyway. There's at least one device in the PR that is a structure, not a machine (crematorium), but is activated via button so now has a wifi receiver.

mwerezak commented:
It would be a good idea to add those checks to the procs added in the PR then.

Checking stat for NOPOWER|BROKEN should be sufficient for most of them, with EMPED thrown in where appropriate.

Other than that, :+1:

Loganbacca commented:
Confirming they all have the appropriate checks.

mwerezak commented on code/game/machinery/buttons.dm:
```diff
@@ -5,14 +5,127 @@
+
+	operating = 1
+	active = !active
+	use_power(5)
```
We use power but we never check to see if we have a source of power :/

Loganbacca replied:
```button/activate()``` is only ever called by ```attack_hand()``` whose parent ```machinery/attack_hand()``` does all the checks: https://github.com/Baystation12/Baystation12/blob/dev/code/game/machinery/machinery.dm#L221

GinjaNinja32 replied:
The parent doing checks doesn't help if you ignore them:
```
/obj/machinery/button/attack_hand(mob/living/user)
	..() // this should likely be 'if(..()) return 1'
	activate(user)
```

Loganbacca replied:
Ah, ok. I didn't realize you had to call it like that to quit out at this level. I'll setup another PR with some mapping changes.

mwerezak commented on code/game/machinery/buttons.dm:
```diff
@@ -5,14 +5,127 @@
+	..()
+	activate(user)
+
+/obj/machinery/button/proc/activate(mob/living/user)
```
I'm having trouble seeing a check for stat here :/
